File: src/openplan/adapters/mesh.py
import json
import os

from openplan.core.ports import MeshBaseline


class MeshAdapter:
    def __init__(self, api_url: str = "", api_key: str = ""):
        self.api_url = api_url
        self.api_key = api_key
        self.cached_baselines: list[MeshBaseline] = []

    async def push_checkpoints(self, checkpoints: list[dict]) -> bool:
        if not self.api_url or not self.api_key:
            return False
        import httpx
        try:
            async with httpx.AsyncClient() as client:
                resp = await client.post(
                    f"{self.api_url}/v1/checkpoints",
                    json={"checkpoints": checkpoints},
                    headers={"Authorization": f"Bearer {self.api_key}"},
                    timeout=10,
                )
                return resp.status_code == 202
        except Exception:
            return False

    async def pull_baselines(self) -> list[MeshBaseline]:
        if not self.api_url or not self.api_key:
            return self.cached_baselines
        import httpx
        try:
            async with httpx.AsyncClient() as client:
                resp = await client.get(
                    f"{self.api_url}/v1/baselines",
                    headers={"Authorization": f"Bearer {self.api_key}"},
                    timeout=10,
                )
                if resp.status_code == 200:
                    data = resp.json()
                    baselines = [
                        MeshBaseline(
                            match_level=b.get("match_level", "action"),
                            action=b.get("action", ""),
                            phase_label_tokens=b.get("phase_label_tokens", ""),
                            avg_cost=b.get("avg_cost", 0),
                            ci_lo=b.get("ci_lo", 0),
                            ci_hi=b.get("ci_hi", 0),
                            sample_count=b.get("sample_count", 0),
                            success_rate=b.get("success_rate", 0),
                        )
                        for b in data.get("baselines", [])
                    ]
                    self.cached_baselines = baselines
                    return baselines
        except Exception:
            pass
        return self.cached_baselines

    def sync_pending(self, conn) -> int:
        """Sync unsynced calibration events. Returns count synced."""
        if not self.api_url or not self.api_key:
            return 0
        pending = conn.execute(
            "SELECT * FROM calibration_events WHERE synced = 0 LIMIT 100"
        ).fetchall()
        if not pending:
            return 0

        checkpoints = []
        for p in pending:
            checkpoints.append({
                "action": p["action"],
                "phase_label_tokens": p["phase_label_tokens"],
                "expected_cost": p["expected_cost"],
                "actual_cost": p["actual_cost"],
                "outcome": p["outcome"],
            })

        import httpx
        try:
            import httpx as _httpx
            resp = _httpx.post(
                f"{self.api_url}/v1/checkpoints",
                json={"checkpoints": checkpoints},
                headers={"Authorization": f"Bearer {self.api_key}"},
                timeout=10,
            )
            if resp.status_code == 202:
                conn.executemany(
                    "UPDATE calibration_events SET synced = 1 WHERE id = ?",
                    [(p["id"],) for p in pending],
                )
                conn.commit()
                return len(pending)
        except Exception:
            pass
        return 0

File: src/adapters/mesh.ts
import type { CalibrationEvent, CostBaseline } from "../core/domain.js";
import { createLogger } from "../core/logger.js";
import type { MeshSync } from "../core/ports.js";
import { BaselinesResponse } from "../core/schemas.js";

const VALID_MATCH_LEVELS = new Set(["exact", "label_keyword", "action"]);

const log = createLogger("mesh");

function outcomeToMesh(outcome: CalibrationEvent["outcome"]): string {
  if (outcome === "completed") return "success";
  if (outcome === "abandoned") return "failure";
  return "partial";
}

export function createMeshSync(meshUrl: string | null, apiKey: string | null): MeshSync {
  const baseUrl = meshUrl ?? "https://api.openplan.cc";

  return {
    async syncCheckpoints(events: CalibrationEvent[]): Promise<boolean> {
      if (!baseUrl) return false;

      try {
        const batch = events.map((e) => ({
          action: e.action,
          phase_label_tokens: e.phaseLabelTokens,
          expected_cost: e.expectedCost,
          actual_cost: e.actualCost,
          outcome: outcomeToMesh(e.outcome),
          session_id: crypto.randomUUID(),
          project_type: e.projectType,
          timestamp: new Date(e.createdAt).getTime() / 1000,
        }));

        const headers: Record<string, string> = {
          "Content-Type": "application/json",
        };
        if (apiKey) headers.Authorization = `Bearer ${apiKey}`;

        const res = await fetch(`${baseUrl}/v1/checkpoints`, {
          method: "POST",
          headers,
          body: JSON.stringify({ events: batch }),
        });

        return res.ok;
      } catch (e) {
        log.warn("Mesh sync failed", e);
        return false;
      }
    },

    async fetchBaselines(): Promise<CostBaseline[] | null> {
      if (!baseUrl) return [];

      try {
        const headers: Record<string, string> = {
          "Content-Type": "application/json",
        };
        if (apiKey) headers.Authorization = `Bearer ${apiKey}`;

        const res = await fetch(`${baseUrl}/v1/baselines`, { headers });
        if (!res.ok) return null;

        const body: unknown = await res.json();
        const parsed = BaselinesResponse.parse(body);
        const rawBaselines = Array.isArray(parsed) ? parsed : parsed.baselines;

        return rawBaselines.map((b) => {
          const rawMatch = typeof b.match_level === "string" ? b.match_level : "";
          const matchLevel: CostBaseline["matchLevel"] = VALID_MATCH_LEVELS.has(rawMatch)
            ? (rawMatch as CostBaseline["matchLevel"])
            : "action";
          return {
            id: crypto.randomUUID(),
            matchLevel,
            action: typeof b.action === "string" ? b.action : "",
            avgCost: typeof b.cost_tokens === "number" ? b.cost_tokens : typeof b.p50 === "number" ? b.p50 : 0,
            ciLo: typeof b.p25 === "number" ? b.p25 : null,
            ciHi: typeof b.p75 === "number" ? b.p75 : null,
            sampleCount: typeof b.sample_count === "number" ? b.sample_count : 0,
            createdAt: new Date().toISOString(),
          };
        });
      } catch (e) {
        log.warn("Baseline fetch failed", e);
        return [];
      }
    },

    async isReachable(): Promise<boolean> {
      if (!baseUrl) return false;

      try {
        const res = await fetch(`${baseUrl}/v1/health`, {
          signal: AbortSignal.timeout(3000),
        });
        return res.ok;
      } catch (e) {
        log.debug("Mesh unreachable", e);
        return false;
      }
    },
  };
}
